### PR TITLE
add metrics for crane-agent

### PR DESCRIPTION
**Pull request by yan234280533**

### CODE REVIEW

**Comment by yan234280533** [comment]
Case1: Ananalye Status and Excuter Status

Before the pressure：
image

**Comment by yan234280533** [comment]
Case2: Health Check
1、 Curl the health check endpoint:
[root@VM-1-50-centos ~]# curl http://172.16.0.33:8081/health-check
OK[root@VM-1-50-centos ~]# curl http://172.16.0.33:8081/health-check
OK[root@VM-1-50-centos ~]# curl http://172.16.0.33:8081/health-check
OK[root@VM-1-50-centos ~]# 

2、Pod status
image

3、Additional delays in the process to simulation of exceptions
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                 node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                 node.kubernetes.io/not-ready:NoExecute op=Exists
                 node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                 node.kubernetes.io/unreachable:NoExecute op=Exists
                 node.kubernetes.io/unschedulable:NoSchedule op=Exists
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  6m11s                default-scheduler  Successfully assigned crane-system/crane-agent-b2xsp to 10.0.1.50
  Normal   Pulled     6m9s                 kubelet            Successfully pulled image "docker.io/gocrane/crane-agent:80e362e" in 1.309292406s
  Normal   Pulling    37s (x2 over 6m10s)  kubelet            Pulling image "docker.io/gocrane/crane-agent:80e362e"
  Warning  Unhealthy  37s (x3 over 57s)    kubelet            Liveness probe failed: HTTP probe failed with statuscode: 500
  Normal   Killing    37s                  kubelet            Container crane-agent failed liveness probe, will be restarted
  Normal   Created    35s (x2 over 6m9s)   kubelet            Created container crane-agent
  Normal   Started    35s (x2 over 6m9s)   kubelet            Started container crane-agent
  Normal   Pulled     35s                  kubelet            Successfully pulled image "docker.io/gocrane/crane-agent:80e362e" in 1.282994704s

Health Check is failed and the container restarted.

**Comment by yan234280533** [comment]
Case2:  Time cost of process

1、Typically time consuming
image

2、Additional delays in the process 
image

**Comment by yan234280533** [comment]
Case4:  Pod evict

The eviction is triggered

image

image

image

File: pkg/ensurance/analyzer/analyzer.go [hunk 1]
```diff
@@ -421,34 +434,51 @@ func (s *AnormalyAnalyzer) getEvictPods(triggered bool, action *ensuranceapi.Avo
...
-func (s *AnormalyAnalyzer) disableSchedulingMerge(acsFiltered []ecache.ActionContext, avoidanceMaps map[string]*ensuranceapi.AvoidanceAction, ae *executor.AvoidanceExecutor) (enableSchedule bool) {
+func (s *AnormalyAnalyzer) disableSchedulingMerge(acsFiltered []ecache.ActionContext, avoidanceMaps map[string]*ensuranceapi.AvoidanceAction, ae *executor.AvoidanceExecutor) bool {
 	var now = time.Now()
-	enableSchedule = false
-	for _, ac := range acsFiltered {
-		if ac.Triggered {
-			enableSchedule = false
-		}
-		if ac.Restored {
+	var enableScheduling, avoidance, restore bool
```
Comment (chenkaiyue) [inline pkg/ensurance/analyzer/analyzer.go:439]: The logic is ok , i test some situations. But still a little confused about these 3 var, what's the difference? maybe we can add some comments.

Comment (yan234280533) [reply pkg/ensurance/analyzer/analyzer.go:443]: done~

File: cmd/crane-agent/app/options/option.go [hunk 2]
```diff
@@ -32,7 +42,10 @@ func (o *Options) Validate() error {
 // AddFlags adds flags to the specified FlagSet.
 func (o *Options) AddFlags(flags *pflag.FlagSet) {
 	flags.StringVar(&o.HostnameOverride, "hostname-override", "", "Which is the name of k8s node be used to filtered.")
-
 	flags.StringVar(&o.RuntimeEndpoint, "runtime-endpoint", "unix:///var/run/dockershim.sock", "The runtime endpoint, default to docker: unix:///var/run/dockershim.sock, containerd: unix:///run/containerd/containerd.sock.")
+	flags.Bool("enable-profiling", false, "Is debug/pprof endpoint enabled, default: false")
+	flags.StringVar(&o.BindAddr, "bind-address", "0.0.0.0:8081", "The address the agent binds to,default: 0.0.0.0:8081.")
```
Comment (mfanjie) [inline cmd/crane-agent/app/options/option.go:47]: can you state what's the purpose of the address binding? for pprof or health check?

File: cmd/crane-agent/app/options/option.go [hunk 3]
```diff
@@ -32,7 +42,10 @@ func (o *Options) Validate() error {
...
+	flags.DurationVar(&o.CollectInterval, "collect-interval", 10*time.Second, "period for the state collector to collect metrics, default: 10s")
```
Comment (mfanjie) [inline cmd/crane-agent/app/options/option.go:48]: Capital letter `Period`?

File: cmd/crane-agent/app/options/option.go [hunk 4]
```diff
@@ -32,7 +42,10 @@ func (o *Options) Validate() error {
...
 	flags.StringArrayVar(&o.Ifaces, "ifaces", []string{"eth0"}, "The network devices to collect metric, use comma to separated, default: eth0")
+	flags.DurationVar(&o.MaxInactivity, "max-inactivity", 5*time.Minute, "Maximum time from last recorded activity before automatic restart, default: 5min")
```
Comment (mfanjie) [inline cmd/crane-agent/app/options/option.go:50]: how user can specify a duration value?

File: pkg/agent/agent.go [hunk 5]
```diff
@@ -2,18 +2,24 @@ package agent
 
 import (
 	"context"
+	"net/http"
+	"time"
 
+	"github.com/gocrane/crane/pkg/metrics"
```
Comment (mfanjie) [inline pkg/agent/agent.go:8]: crane import should be in other section

File: pkg/agent/agent.go [hunk 6]
```diff
@@ -2,18 +2,24 @@ package agent
...
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/component-base/metrics/legacyregistry"
 	"k8s.io/klog/v2"
 
 	ensuranceapi "github.com/gocrane/api/ensurance/v1alpha1"
 	craneclientset "github.com/gocrane/api/pkg/generated/clientset/versioned"
 	"github.com/gocrane/api/pkg/generated/informers/externalversions/ensurance/v1alpha1"
-
+	"github.com/gocrane/crane/cmd/crane-agent/app/options"
```
Comment (mfanjie) [inline pkg/agent/agent.go:22]: why source code in pkg import pkg cmd? we should make sure import to be one direction.

File: pkg/ensurance/analyzer/analyzer.go [hunk 7]
```diff
@@ -114,7 +116,14 @@ func (s *AnormalyAnalyzer) Run(stop <-chan struct{}) {
 		for {
 			select {
 			case state := <-s.stateChann:
-				go s.Analyze(state)
+				{
```
Comment (mfanjie) [inline pkg/ensurance/analyzer/analyzer.go:119]: why need {} here?

File: pkg/ensurance/collector/collector.go [hunk 8]
```diff
@@ -45,12 +53,19 @@ func (s *StateCollector) Name() string {
...
 func (s *StateCollector) Run(stop <-chan struct{}) {
 	go func() {
-		updateTicker := time.NewTicker(10 * time.Second)
+		updateTicker := time.NewTicker(s.collectInterval)
 		defer updateTicker.Stop()
 		for {
 			select {
 			case <-updateTicker.C:
-				s.UpdateCollectors()
+				{
```
Comment (mfanjie) [inline pkg/ensurance/collector/collector.go:61]: same as above, do we need {} here?

File: pkg/metrics/ensuarance.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+)
+
+var (
+	/**** Metrics related to crane-agent execution ****/
+	lastActivity = k8smetrics.NewGaugeVec(
+		&k8smetrics.GaugeOpts{
+			Namespace:      CraneNamespace,
+			Subsystem:      CraneAgentSubsystem,
+			Name:           LastActivity,
+			Help:           "Last time certain part of crane logic executed.",
```
Comment (mfanjie) [inline pkg/metrics/ensuarance.go:72]: what logic? can you make it more specific?

Comment (yan234280533) [reply pkg/metrics/ensuarance.go:72]: done~

**Pull request merged**